srtcore/api.cpp

```diff
@@ -3027,26 +3027,34 @@ int CUDT::removeSocketFromGroup(SRTSOCKET socket)
     if (!s->m_IncludedGroup)
         return APIError(MJ_NOTSUP, MN_INVAL, 0);
 
-    ScopedLock grd (s->m_ControlLock);
     s->removeFromGroup(false);
     return 0;
 }
 
 void CUDTSocket::removeFromGroup(bool broken)
 {
-    m_IncludedGroup->remove(m_SocketID);
-    if (broken)
+    CUDTGroup* pg = 0;
     {
-        // Activate the SRT_EPOLL_UPDATE event on the group
-        // if it was because of a socket that was earlier connected
-        // and became broken. This is not to be sent in case when
-        // it is a failure during connection, or the socket was
-        // explicitly removed from the group.
-        m_IncludedGroup->activateUpdateEvent();
+        ScopedLock grd (m_ControlLock);
+        pg = m_IncludedGroup;
+        m_IncludedIter = CUDTGroup::gli_NULL();
+        m_IncludedGroup = NULL;
     }
 
-    m_IncludedIter = CUDTGroup::gli_NULL();
-    m_IncludedGroup = NULL;
+    // Another facility could have deleted it in the meantime.
+    if (pg)
+    {
+        pg->remove(m_SocketID);
+        if (broken)
+        {
+            // Activate the SRT_EPOLL_UPDATE event on the group
+            // if it was because of a socket that was earlier connected
+            // and became broken. This is not to be sent in case when
+            // it is a failure during connection, or the socket was
+            // explicitly removed from the group.
+            pg->activateUpdateEvent();
+        }
+    }
 }
```

Comment:
Reduced nesting:

```c++
void CUDTSocket::removeFromGroup(bool broken)
{
    // First lock access and clear pointers
    enterCS(m_ControlLock);
    CUDTGroup* pg = m_IncludedGroup;
    m_IncludedIter = CUDTGroup::gli_NULL();
    m_IncludedGroup = NULL;
    leaveCS(m_ControlLock);

    if (!pg)
        return;

    pg->remove(m_SocketID);
    if (broken)
    {
        // Activate the SRT_EPOLL_UPDATE event on the group
        // if it was because of a socket that was earlier connected
        // and became broken. This is not to be sent in case when
        // it is a failure during connection, or the socket was
        // explicitly removed from the group.
        pg->activateUpdateEvent();
    }
}
```

Reply:
This reduces clarity as well. A critical section inside braces marks it clearer than the pair of enter/leave calls. The return on no pg looks good.

Reply:
Wherever there is a code block without `if`s, `while`s, etc., it means there should have been a function.


```c++
// A sample name, not pretending to be good
CUDTGroup* CUDTSocket::clearGroupMembership()
{
    ScopedLock grd (m_ControlLock);
    CUDTGroup* pg = m_IncludedGroup;
    m_IncludedIter = CUDTGroup::gli_NULL();
    m_IncludedGroup = NULL;
    return pg;
}

void CUDTSocket::removeFromGroup(bool broken)
{
    // First lock access and clear pointers
    CUDTGroup* pg = clearGroupMembership();
    if (!pg)
        return;

    pg->remove(m_SocketID);
    if (broken)
    {
        // Activate the SRT_EPOLL_UPDATE event on the group
        // if it was because of a socket that was earlier connected
        // and became broken. This is not to be sent in case when
        // it is a failure during connection, or the socket was
        // explicitly removed from the group.
        pg->activateUpdateEvent();
    }
}
```